toolchain/check/generic_region_stack.h

```diff
@@ -0,0 +1,74 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_GENERIC_REGION_STACK_H_
+#define CARBON_TOOLCHAIN_CHECK_GENERIC_REGION_STACK_H_
+
+#include "llvm/ADT/BitmaskEnum.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::Check {
+
+LLVM_ENABLE_BITMASK_ENUMS_IN_NAMESPACE();
+
+// A stack of enclosing regions that might be declaring or defining a generic
+// entity. In such a region, we track the generic constructs that are used, such
+// as symbolic constants and types, and instructions that depend on a template
+// parameter.
+//
+// TODO: For now we're just tracking symbolic constants.
+//
+// We split a generic into two regions -- declaration and definition -- because
+// these are in general introduced separately, and substituted into separately.
+// For example, for `class C(T:! type, N:! T) { var x: T; }`, a use such as
+// `C(i32, 0)*` substitutes into just the declaration, whereas a use such as
+// `var x: C(i32, 0) = {.x = 0};` also substitutes into the definition.
+class GenericRegionStack {
+ public:
+  // Ways in which an instruction can depend on a generic parameter.
+  enum class DependencyKind : int8_t {
+    None = 0x0,
+    // The type of the instruction depends on a checked generic parameter.
+    SymbolicType = 0x1,
+    // The constant value of the instruction depends on a checked generic
+    // parameter.
+    SymbolicConstant = 0x2,
+    Template = 0x4,
+    LLVM_MARK_AS_BITMASK_ENUM(/*LargestValue=*/Template)
+  };
+
+  // An instruction that depends on a generic parameter in some way.
+  struct DependentInst {
+    SemIR::InstId inst_id;
+    DependencyKind kind;
+  };
+
+  // Pushes a region that might be declaring or defining a generic.
+  auto Push() -> void;
+
+  // Pops a generic region.
+  auto Pop() -> void;
+
+  // Adds an instruction to the list of instructions in the current region that
+  // in some way depend on a generic parameter.
+  auto AddDependentInst(DependentInst inst) -> void;
+
+  // Returns the list of dependent instructions in the current generic region.
+  auto PeekDependentInsts() -> llvm::ArrayRef<DependentInst>;
+
+ private:
+  struct RegionInfo {
+    int32_t first_dependent_inst;
+  };
+
+  // The current set of enclosing generic regions.
+  llvm::SmallVector<RegionInfo> regions_;
+
+  // List of symbolic constants used in any of the enclosing generic regions.
+  llvm::SmallVector<DependentInst> dependent_insts_;
```

Comment:
If I understand correctly, this tracks the insts as a single vector, using RegionInfo to slice up sections. It looks like there's no use of spanning regions, is that correct? Why this approach versus a vector of regions with vector<DependentInst> as a per-region member? I can see there may be reasons, maybe you'll want the overlapping regions for member class types or similar, but a comment might help clarify.

I think the relationship between regions_ and dependent_insts_ would be helpful to talk about a little in particular, as the structure manipulation requires a bit of reading to understand.

Reply:
Added a comment. The motivation for this approach is denser storage and avoiding reallocations.

[FWIW, originally I had a second reason: for cases like

```carbon
class A(T:! type) {
  fn F(a: T) -> T;
}
```

... I was not introducing a separate generic region for the declaration of `F`, and instead was using the definition region of `A` to track the references to `T` inside the declaration of `A(T).F`. But when you start processing `F`, you don't know whether it has its own generic region or not, so the idea was that I'd track the dependent instructions in a flat list, and if `F` turns out to not be a generic function, we can just pop it from the list of regions and keep its instructions in the enclosing region.

But that turns out to not really work so well. It adds complexity in various places, and we end up wanting a separate generic region for the definition of `F` anyway, so it seems better to say that in this case, `F` is a generic function that just happens to have no more generic parameters than `A` does, so that's the direction I'm now pursuing.]